src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs

```diff
@@ -167,61 +192,63 @@ protected static Package FromFlatOpcDocumentCore(XDocument document, Package pac
+                throw new ArgumentException(ExceptionMessages.RootElementIsNull, nameof(document));
+            }
+
             // Add all parts (but not relationships).
-            foreach (var xmlPart in document.Root
+            foreach (XElement xmlPart in document.Root
                 .Elements()
                 .Where(p =>
-                    (string)p.Attribute(pkg + "contentType") !=
-                        "application/vnd.openxmlformats-package.relationships+xml"))
+                    (string) p.Attribute(Pkg + "contentType") !=
+                    "application/vnd.openxmlformats-package.relationships+xml"))
             {
-                string name = (string)xmlPart.Attribute(pkg + "name");
-                string contentType = (string)xmlPart.Attribute(pkg + "contentType");
-                if (contentType.EndsWith("xml", StringComparison.Ordinal))
+                var name = (string) xmlPart.Attribute(Pkg + "name");
+                var contentType = (string) xmlPart.Attribute(Pkg + "contentType");
+                if (contentType.EndsWith("xml"))
                 {
-                    Uri uri = new Uri(name, UriKind.Relative);
+                    var uri = new Uri(name, UriKind.Relative);
                     PackagePart part = package.CreatePart(uri, contentType, CompressionOption.SuperFast);
                     using (Stream stream = part.GetStream(FileMode.Create))
                     using (XmlWriter xmlWriter = XmlWriter.Create(stream))
                     {
-                        xmlPart.Element(pkg + "xmlData")
+                        xmlPart.Elements(Pkg + "xmlData")
                             .Elements()
                             .First()
                             .WriteTo(xmlWriter);
                     }
                 }
                 else
                 {
-                    Uri uri = new Uri(name, UriKind.Relative);
+                    var uri = new Uri(name, UriKind.Relative);
                     PackagePart part = package.CreatePart(uri, contentType, CompressionOption.SuperFast);
                     using (Stream stream = part.GetStream(FileMode.Create))
-                    using (BinaryWriter binaryWriter = new BinaryWriter(stream))
+                    using (var binaryWriter = new BinaryWriter(stream))
                     {
-                        string base64StringInChunks = (string)xmlPart.Element(pkg + "binaryData");
-                        char[] base64CharArray = base64StringInChunks
-                            .Where(c => c != '\r' && c != '\n').ToArray();
-                        byte[] byteArray =
-                            System.Convert.FromBase64CharArray(
-                                base64CharArray, 0, base64CharArray.Length);
+                        var base64StringInChunks = (string) xmlPart.Element(Pkg + "binaryData");
+                        char[] base64CharArray = base64StringInChunks.Where(c => c != '\r' && c != '\n').ToArray();
```

Comment:
Can you move this out to a function as well?

Reply:
Done. And you got me started on a more complete refactoring of the other methods as well.